cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -238,104 +238,75 @@ const updateHostedZoneRecords = async function (
-// Therefore, we check if there is more than one certificate using the record and only delete
-// if there is no other certificate using the record.
 const deleteHostedZoneRecords = async function (
-  options,
-  certDomain,
+  oldCertOptions,
+  certArn,
+  defaultDomain,
   envRoute53,
   appRoute53,
   acm,
   envHostedZoneId
 ) {
-  let isLegacyCert = false;
-  // Legacy cert only has two DomainValidationOptions:
-  // `${envName}.${appName}.${domainName}` and `*.${envName}.${appName}.${domainName}`
-  if (options.length <= 2) {
-    isLegacyCert = true;
-  }
-
-  const certsWithEnvDomain = await numOfGeneratedCertificates(acm, certDomain);
-  const isLastOne = certsWithEnvDomain === 1;
-
-  const newOptions = [];
-  switch (`${isLegacyCert}|${isLastOne}`) {
-    case `true|true`:
-      // If it is a legacy cert and it is the last Copilot cert,
-      // we'll go ahead to remove the validation CNAME record in env hosted zone.
-      newOptions.push(...options);
-      break;
-    case `true|false`:
-      // If it is a legacy cert but it is not the last Copilot cert,
-      // we'll do nothing since the new Copilot cert needs validation CNAME record
-      // in the env hosted zone.
-      break;
-    case `false|true`:
-      // If it is not a legacy cert and it is the last Copilot cert,
-      // we'll remove all validation CNAME records in env/app/root hosted zone.
-      newOptions.push(...options);
-      break;
-    case `false|false`:
-      // If it is not a legacy cert and it is not the last Copilot cert,
-      // we'll remove validation CNAME records only for app and root hosted zone,
-      // since the legacy cert still needs the validation record in the env hosted zone.
-      for (const option of options) {
-        if (option.DomainName === certDomain || option.DomainName === `*.${certDomain}`) {
-          continue;
-        }
-        newOptions.push(option);
+  let listCertificatesInput = {};
+  let newCertOptions = [];
+  // Make sure if there is a new certificate, we don't delete any DNS validation records
+  // that are used by the new certificate.
+  while (true) {
+    const listCertResp = await acm
+      .listCertificates(listCertificatesInput)
+      .promise();
+    for (const certSummary of listCertResp.CertificateSummaryList || []) {
+      if (
+        certSummary.DomainName !== defaultDomain ||
+        certSummary.CertificateArn === certArn
+      ) {
+        // Skip if it is not the new certificate.
+        continue;
       }
+      // There exists another certificate created by Copilot which has the updated alias fields as SANs.
+      // We don't want to delete any validation records associated with the new certificate.
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: certSummary.CertificateArn,
+        })
+        .promise();
+      newCertOptions = Certificate.DomainValidationOptions || [];
```

Comment:
We intend to break out of the larger `while` loop right?

Reply:
yeah cause that means we found it

Reply:
This probably doesn't matter since there should be only one such cert. But it seems lie we are only breaking out of the for-loop, not the while loop 🤔

Reply:
yeah that's true. OK done fixing it.